### PR TITLE
mapping bugs; fixes #54

### LINKED ISSUE #54: Mapping and mapping configuration small bugs
### Prerequisites

- [x] I have written a descriptive issue title
- [x] I have verified that I am running the latest version of the DEH-CATIA
- [x] I have searched [open](https://github.com/RHEAGROUP/DEH-CATIA/issues) and [closed](https://github.com/RHEAGROUP/DEH-CATIA/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### Description

- [x] Option and State dependency not handled correctly in mapping rule from Catia to the Hub
- [ ] #55
- [x] ParameterOverride don't override an existing parameter possibly. Results in "More than one parameter returned" exception from the Hub
- [x] MappingConfiguration Containement exception when saving new one after a failed transfer
- [x] Mapping configuration exception when trying to deserialized

### System Configuration

- DEHPCommon version:
- Environment (Operating system, version and so on):
- .NET Framework version:
- Additional information:

### PR DESCRIPTION
**Pull request by nathanatrhea**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/RHEAGROUP/DEH-CATIA/pulls) open
- [x] I have verified that I am following the DEH-CATIA [code style guidelines](https://raw.githubusercontent.com/RHEAGROUP/DEH-CATIA/master/.github/CONTRIBUTING.md)
- [x] I have provided test coverage for my change (where applicable)

### Description

;fixes #54
- [x] ParameterOverride parameter reference problem fixed
- [x] MappingConfiguration Containement exception when saving new one after a failed transfer
- [x] Mapping configuration exception when trying to deserialized
- [x] added proper support for state and option dependent parameter

### CODE REVIEW

**Comment by nathanatrhea** [comment]
image

File: DEHCATIA/MappingRules/CatiaProductToElementRule.cs [hunk 1]
```diff
@@ -560,30 +557,112 @@ private ParameterOverride GetParameterOverride(ParameterType parameterType, Elem
...
+                Computed = new ValueArray<string>(initializationCollection),
+                Formula = new ValueArray<string>(initializationCollection),
+                Manual = new ValueArray<string>(initializationCollection),
+                Reference = new ValueArray<string>(initializationCollection),
+                Published = new ValueArray<string>(initializationCollection)
+            };
+        }
+
+
+        private ParameterValueSet CreateParameterValueSet(IList<string> initializationCollection, Option option = null,
```
Comment (alexatrhea) [inline DEHCATIA/MappingRules/CatiaProductToElementRule.cs:651]: docs

File: DEHCATIA/MappingRules/CatiaProductToElementRule.cs [hunk 2]
```diff
@@ -560,30 +557,112 @@ private ParameterOverride GetParameterOverride(ParameterType parameterType, Elem
...
+            }
+            else
+            {
+                valueSets.Add(this.CreateValueSet<TValueSet>(parameterToOverride, initializationCollection));
+            }
+
+            return valueSets;
+        }
+
+        private TValueSet CreateValueSet<TValueSet>(Parameter parameter, IList<string> initializationCollection, Option option = null,
```
Comment (alexatrhea) [inline DEHCATIA/MappingRules/CatiaProductToElementRule.cs:612]: Docs

File: DEHCATIA/MappingRules/CatiaProductToElementRule.cs [hunk 3]
```diff
@@ -560,30 +557,112 @@ private ParameterOverride GetParameterOverride(ParameterType parameterType, Elem
...
 
+                parameter.ValueSet.AddRange(this.CreateValueSets<ParameterOverrideValueSet>(parameterToOverride, initializationCollection));
+
                 elementUsage.ParameterOverride.Add(parameter);
             }
 
             return parameter;
         }
 
+        private IEnumerable<TValueSet> CreateValueSets<TValueSet>(Parameter parameterToOverride, IList<string> initializationCollection) where TValueSet : ParameterValueSetBase, new()
```
Comment (alexatrhea) [inline DEHCATIA/MappingRules/CatiaProductToElementRule.cs:576]: Docs

File: DEHCATIA/ViewModels/Rows/MappedElementRowViewModel.cs [hunk 4]
```diff
@@ -120,53 +168,92 @@ public MappedElementRowViewModel()
...
+                this.HasHubElementAnyStateDependentParameter = false;
             }
             else
             {
-                if (element is UsageRowViewModel && this.HubElement is ElementUsage
-                    || element is DefinitionRowViewModel && this.HubElement is ElementDefinition)
+                this.HasHubElementAnyOptionDependentParameter = element.HasAnyDependencyOnOption();
+                this.HasHubElementAnyStateDependentParameter = element.HasAnyDependencyOnActualFiniteState();
+
+                //(this.HasHubElementAnyOptionDependentParameter, this.HasHubElementAnyStateDependentParameter) = element switch
```
Comment (alexatrhea) [inline DEHCATIA/ViewModels/Rows/MappedElementRowViewModel.cs:196]: Commented code

**Review by alexatrhea — CHANGES_REQUESTED** [review]
Some small things to fix

**Review by alexatrhea — APPROVED** [review]

**Review by antoineatrhea — APPROVED** [review]

**Pull request merged**